index/scorch/scorch.go

```diff
@@ -54,6 +54,7 @@ type Scorch struct {
 	rootLock             sync.RWMutex
 	root                 *IndexSnapshot // holds 1 ref-count on the root
 	rootPersisted        []chan error   // closed when root is persisted
+	callbacks            []index.BatchCallback
```

Comment:
Hmmm, still thinking after reading marty's comments on global versus local callbacks...

but, this stands out to me as worrisome now, and I should have seen it earlier (my apologies).  I'm thinking this tracking of callbacks ought to be on the snapshot, not on the Scorch instance, as putting it on the Scorch instance feels more 'global' to me now.

Reply:
I think we should separate between 2 things:
1. the APIs to be exposed
2. the implementation, and where to hold each struct/func

Regarding the APIs, i think that a more explicit approach makes it easier for a developer to use.
Regarding the implementation of `this` PR i used the same mechanism of `persistent` channel
For simplicity i would leave the implementation this way, and see how it works
The API will should stay this way anyhow ...